packages/angular/build/src/utils/index-file/auto-csp.ts

```diff
@@ -0,0 +1,322 @@
+ */
+function shouldDynamicallyLoadScriptTagBasedOnType(scriptType: string | undefined): boolean {
+  return (
+    scriptType === undefined ||
+    scriptType === '' ||
+    scriptType === 'module' ||
+    isJavascriptMimeType(scriptType)
+  );
+}
+
+/**
+ * Calculates a CSP compatible hash of an inline script.
+ * @param scriptText Text between opening and closing script tag. Has to
+ *     include whitespaces and newlines!
+ * @returns The hash of the text formatted appropriately for CSP.
+ */
+export function hashScriptText(scriptText: string): string {
+  const hash = crypto.createHash(HASH_FUNCTION).update(scriptText, 'utf-8').digest('base64');
+  return `'${HASH_FUNCTION}-${hash}'`;
+}
+
+/**
+ * Generates the dynamic loading script and puts it in the rewriter and adds the hash of the dynamic
+ * loader script to the collection of hashes to add to the <meta> tag CSP.
+ *
+ * @param scriptContent The current streak of <script src="..."> tags
+ * @param hashes The array of hashes to include in the final CSP
+ * @param rewriter Where to emit tags to
+ */
+function emitLoaderScript(
+  scriptContent: SrcScriptTag[],
+  hashes: string[],
+  rewriter: RewritingStream,
+) {
+  const loaderScript = createLoaderScript(scriptContent);
+  hashes.push(hashScriptText(loaderScript));
+  rewriter.emitRaw(`<script>${loaderScript}</script>`);
+}
+
+/**
+ * Finds all `<script>` tags and creates a dynamic script loading block for consecutive `<script>` with `src` attributes.
+ * Hashes all scripts, both inline and generated dynamic script loading blocks.
+ * Inserts a `<meta>` tag at the end of the `<head>` of the document with the generated hash-based CSP.
+ *
+ * @param html Markup that should be processed.
+ * @returns The transformed HTML that contains the `<meta>` tag CSP and dynamic loader scripts.
+ */
+export async function autoCsp(html: string): Promise<string> {
+  const { rewriter, transformedContent } = await htmlRewritingStream(html);
+
+  let openedScriptTag: StartTag | undefined = undefined;
+  let scriptContent: SrcScriptTag[] = [];
+  let hashes: string[] = [];
+
+  rewriter.on('startTag', (tag, html) => {
+    if (tag.tagName === 'script') {
+      openedScriptTag = tag;
+      const src = getScriptAttributeValue(tag, 'src');
+
+      if (src) {
+        // If there are any interesting attributes, note them down.
+        const scriptType = getScriptAttributeValue(tag, 'type');
+        if (shouldDynamicallyLoadScriptTagBasedOnType(scriptType)) {
+          scriptContent.push({
+            scriptType: 'src',
+            src: src,
+            type: scriptType,
+            async: !(getScriptAttributeValue(tag, 'async') === undefined),
+            defer: !(getScriptAttributeValue(tag, 'defer') === undefined),
+          });
+          return; // Skip writing my script tag until we've read it all.
+        }
+      }
+    }
+    // We are encountering the first start tag that's not <script src="..."> after a string of
```

Comment:
I think we want the loader script to be written before a `<script>console.log(...)</script>`, if the previous tags were a string of `<script src=...>`, right?

This is captured in the unit test that does

```html
<script>console.log('foo');</script>
<script src="./main.js"></script>
<script src="./main2.js"></script>
<script>console.log('bar');</script>
<script src="./main3.js"></script>
<script src="./main4.js"></script>
<div>Some text </div>
```

Which I thought should be

```html
<script>console.log('foo');</script>
<script>...Loader script for main.js and main2.js</script>
<script>console.log('bar');</script>
<script>...Loader script for main3.js and main4.js</script>
<div>Some text </div>
```

As per a script with a malformed type (i.e. `<script type="not/js" ...>`), the current behavior is "leave it alone", since it's not guaranteed to run anyway according to the spec? If we want to be more forgiving, we can always rewrite it into a loader script if it also contains a `src` attribute, but I thought at least leaving it alone was preferred to silently stripping it from the HTML. (Regardless, we need the logic to fall through to `rewriter.emitStartTag` as written to preserve this behavior.)

Reply:
Comment was updated to call out that we stop on `<script>` tags with no `src`! Thanks!